src/libraries/System.Private.CoreLib/src/System/WeakReference.T.cs

```diff
@@ -15,23 +24,64 @@ namespace System
     public sealed partial class WeakReference<T> : ISerializable
         where T : class?
     {
-        // If you fix bugs here, please fix them in WeakReference<T> at the same time.
+        // If you fix bugs here, please fix them in WeakReference at the same time.
 
         // Most methods using the handle should use GC.KeepAlive(this) to avoid potential handle recycling
         // attacks (i.e. if the WeakReference instance is finalized away underneath you when you're still
         // handling a cached value of the handle then the handle could be freed and reused).
 
-#if !CORECLR
-        // the handle field is effectively readonly until the object is finalized.
-        private IntPtr _handleAndKind;
+        private nint_t _handleAndKind;
+
+        // the lowermost 2 bits are reserved for storing additional info about the handle
+        // we can use these bits because handle is at least 32bit aligned
+        // we also reserve the sign bit
+#if CORECLR
+        private const nint_t HandleTagBits = nint_t.MinValue | 3;
```

Comment:
This is an odd asymmetry. The native and managed should both use the same logic to set the higher order bit. Let's make them both use this approach of the unsigned approach. Personally I think the unsigned approach is clearer since few think about the bits layout for the minimum value of an integer.